rfc/rfc-0014-change-tracking.md

```diff
@@ -0,0 +1,46 @@
+Motivation
+Context
+At the core of the architecture of cartographer is the concept that a supply chain is the orchestration of objects and their controllers.  
+
+A supply chain, as the name suggests, chains together a set of objects defining how the (status) fields of one object feeds into the spec (and sometimes data) fields of another.  Thus creating an ordered chain of interacting objects.  Because controllers continuously reconcile their objects towards a desired state.  A supply chain is, therefore, able to choreograph an otherwise set of independent objects (and their controllers).
+
+A workload acts as input into this supply chain seeding initial values into an object to make the first controller do work.  The supply chain manages this workload input data and outputs from each object as it propagates through the supply chain.
+
+It is important to note however, that a workload can have multiple inputs and input into several objects in the supply chain all at the same time.  As a result when a workload changes, it may cause several controllers to do work simultaneously.  
+
+It is also important to note that a controller may also do work outside of the work that the supply chain is choreographing.  Kpack images, for example, are often choreographed as part of a supply chain.  But the kpack controller may also do work in response to a base OS image update.  Nothing to do with the supply chain but impacting it none-the-less.  As a result the supply chain may be triggered part way through by this input.
+Problem
+Whilst, on the one hand, the behavior described above is virtuous and generally beneficial to automated outer loop workflows.  On the other hand, this behavior can be problematic for more user-centric, imperative workflows, such as those often found in the inner loop.  Those of debugging and live update.  
+
+Providing a real world example.  Given a supply chain that has several inputs, let’s say source code image and a debug flag.  When a developer applies a workload after changing their source code and turning on debug.  Several services in the supply chain may trigger at the same time and these inputs will traverse through the supply chain at different times.  But the developer, initiating the debug session will want to wait for that specific source code change to arrive “in cluster” in an image prepped for debugging before attaching their debugger.  So, it is important to know when both inputs have fully traversed the supply chain.
+
+## Possible Solutions
+As the cartographer is already generally aware of workload inputs and outputs (artifacts) and how those artifacts from one component (service) are inputs into another.  It should, therefore, be possible to output, as additional status, the relationship between these artifacts and the service that produced it.
+
+For example:
+
+```
+status:
+   outputs:
+   - type: image
```

Comment:
What happens when another downstream outputs gets created with this image (e.g. config)? Are we going to have two separate output trees that contain this information?

```yaml
status:
   outputs:
   - type: image
     value: abc123
     passed: [image-builder]
     from:
     - type: source
       value: def567
       passed: [workload, source-cloner, unit-tester]
   - type: config
     value:  shasum(some-giant-config)
     passed: [config-builder]
     from:
     - type: image   <# ========= does this whole tree repeat again?
       value: abc123
       passed: [image-builder]
       from:
       - type: source
         value: def567
         passed: [workload, source-cloner, unit-tester]
```

Reply:
Maybe the nested entries should just be refs to the top-level. E.g.,
```yaml
status:
   outputs:
   - type: source
      value: def567
      passed: [workload, source-cloner, unit-tester]

   - type: image
     value: abc123
     passed: [image-builder]
     from:
     - type: source
       valueRef: def567

   - type: config
     value:  shasum(some-giant-config)
     passed: [config-builder]
     from:
     - type: image
       valueRef: abc123
```

Reply:
@sclevine a source produces two values, `url` and `revision` so we probably want to track both. How would a `valueRef` work for cases like that?

Reply:
Might need IDs/UUIDs or something similar

Reply:
+1 on IDs for each entry (perhaps running a hash function on the value? given that we have the values themselves once per output, we can always find the pre-image of the hashing result if we want need to surface the underlying content of the hash) 

I wonder if `workload` itself could be considered an """output"""  which could be referenced like any other value so that we can always point back to the workload as it was at such point in time 🤔

Reply:
> perhaps running a hash function on the value

💯  Maybe sha256 of ordered JSON?

> I wonder if workload itself could be considered an """output""" which could be referenced like any other value so that we can always point back to the workload as it was at such point in time 🤔

This could get complicated, because it isn't possible to guarantee that all related artifacts are generated from the same Workload spec. We could store revisions of the Workload spec as artifacts, and ref them as inputs of every non-Workload artifact. But I'm not sure if that's worth it, given that we aren't going to be able to capture every input with Cartographer anyways. Might be better to rely on Tekton Chains for Tekton and SBoM for Kpack images for provenance.

Reply:
Updated suggestion:
```yaml
  artifacts:
  - source: # oneOf(source,image,config)
      id: <SHA:string>
      uri: <:string>
      revision: <:string>
      passed: <resource-names:[]string>
      from:
      - id: <SHA:string>
  - image: # oneOf(source,image,config)
      id: <SHA:string>
      digest: <"sha256:"digest:string>
      passed: <resource-names:[]string>
      from:
      - id: <SHA:string>
  - config: # oneOf(source,image,config)
      id: <SHA:string>
      config: <:string>
      passed: <resource-names:[]string>
      from:
      - id: <SHA:string>
```

Reply:
From the [Nov 22 Office Hours](https://docs.google.com/document/d/1ImIh7qBrOLOvGMCzY6AURhE-a68IE9_EbCf0g5s18vc/edit):
Issue #378 points out that it is useful to note not only the component name that has yielded an artifact, but also the object that artifact has been read from. E.g.
```yaml
  artifacts:
  - source: # oneOf(source,image,config)
      id: <SHA:string>
      uri: <:string>
      revision: <:string>
      passed: <resource-names:[]string>
      from:
      - id: <SHA:string>
        kind: <:string>
        apiVersion: <:string>
        name: <:string>
        namespace: <:string>
        resourceVersion: <:string>
```

Reply:
Clarification:
```yaml
  artifacts:
  - source: # oneOf(source,image,config)
      id: <SHA:string>
      uri: <:string>
      revision: <:string>
      passed:
      - resource-name: <:string>
        kind: <:string>
        apiVersion: <:string>
        name: <:string>
        namespace: <:string>
        resourceVersion: <:string>
      from:
      - id: <SHA:string>
```
first item in the list of passed should be the resource that originally created the artifact.